File: frontend/src/api/auth/userSettings.ts
import { API_BACKEND_URL } from "../../config";
import { encryptPassword } from "../../utils/encryptPassword";

export const updateFullName = async (fullName: string) => {
  const token = localStorage.getItem("token");
  if (!token) throw new Error("⛔ Токен не найден");

  const response = await fetch(`${API_BACKEND_URL}/users/update-fullname`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ full_name: fullName }),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.message || "❌ Ошибка обновления ФИО");
  }

  return await response.json();
};

export const updateAvatar = async (avatarBase64: string) => {
  const token = localStorage.getItem("token");
  if (!token) throw new Error("Токен не найден");

  const response = await fetch(`${API_BACKEND_URL}/users/update-avatar`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ avatar: avatarBase64 }),
  });

  const data = await response.json();

  if (!response.ok) {
    console.error("Ошибка при обновлении аватара:", data);
    throw new Error(data.message || "Ошибка обновления аватара");
  }

  if (!data?.data?.avatar) {
    throw new Error("Ответ не содержит URL аватара");
  }

  return data;
};

export const changePassword = async (newPassword: string) => {
  try {
    const token = localStorage.getItem("token");
    if (!token) throw new Error("Токен не найден");

    const encryptedPassword = encryptPassword(newPassword);

    const response = await fetch(`${API_BACKEND_URL}/users/change-password`, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ password: encryptedPassword }),
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData?.message || "Ошибка при смене пароля");
    }

    const responseData = await response.json();
    return responseData;
  } catch (error) {
    console.error("Ошибка при смене пароля:", error);
    throw error;
  }
};

export const subscribeTelegram = async (username: string, telegramId: string) => {
  const response = await fetch(`${API_BACKEND_URL}/users/subscribe`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ username, telegram_id: telegramId }),
  });

  const data = await response.json();

  if (!response.ok) {
    throw new Error(data.message || "Ошибка привязки Telegram");
  }

  return data;
};

export const unlinkTelegram = async (username: string) => {
  const response = await fetch(`${API_BACKEND_URL}/users/unlink`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ username }),
  });

  const data = await response.json();

  if (!response.ok) {
    throw new Error(data.message || "Ошибка отвязки Telegram");
  }

  return data;
};

export const getProfile = async () => {
  const token = localStorage.getItem("token");
  if (!token) throw new Error("Токен не найден");

  const response = await fetch(`${API_BACKEND_URL}/users/profile`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  const data = await response.json();

  if (!response.ok) {
    console.error("Ошибка при получении профиля:", data);
    throw new Error(data.message || "Ошибка получения профиля");
  }

  return data;
};

File: backend/app/routes/users/handlers/user_settings.py
from fastapi import APIRouter, Request, Depends
from sqlalchemy.orm import Session
from app.auth import get_db, get_current_user, get_password_hash
from app.models import User
from app.responses import (
    success_response,
    bad_request_response,
    deleted_response,
    internal_error_response
)
from app.utils.crypto import decrypt_password
from loguru import logger

router = APIRouter()


@router.get("/profile")
def get_profile(user: User = Depends(get_current_user)):
    try:
        return success_response({
            "username": user.username,
            "telegram_id": user.telegram_id,
            "full_name": user.full_name,
            "avatar": user.avatar_url
        }, "Профиль пользователя получен")
    except Exception as e:
        logger.error(f"[GET /profile] Ошибка: {e}")
        return internal_error_response("Не удалось получить профиль пользователя")


@router.patch("/update-fullname")
def update_full_name(data: dict, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    full_name = data.get("full_name")
    if not full_name:
        return bad_request_response("Поле full_name обязательно")

    try:
        user.full_name = full_name
        db.commit()
        return success_response({}, "ФИО успешно обновлено")
    except Exception as e:
        logger.error(f"[PATCH /update-fullname] Ошибка: {e}")
        return internal_error_response("Не удалось обновить ФИО")


@router.patch("/update-avatar")
def update_avatar(data: dict, db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    avatar_base64 = data.get("avatar")
    if not avatar_base64:
        return bad_request_response("Поле avatar обязательно")

    try:
        user.avatar_url = avatar_base64
        db.commit()
        return success_response(
            {"avatar": user.avatar_url},
            "Аватар успешно обновлён"
        )
    except Exception as e:
        logger.error(f"[PATCH /update-avatar] Ошибка: {e}")
        return internal_error_response("Не удалось обновить аватар")


@router.patch("/change-password")
async def change_password(
    request: Request,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user)
):
    try:
        data = await request.json()
        encrypted_password = data.get("password")

        if not encrypted_password:
            return bad_request_response("Поле 'password' обязательно")

        try:
            decrypted_password = decrypt_password(encrypted_password)
        except Exception as e:
            logger.error(f"[change-password] Ошибка расшифровки пароля: {e}")
            return bad_request_response("Некорректный зашифрованный пароль")

        hashed_password = get_password_hash(decrypted_password)
        user.password_hash = hashed_password
        db.commit()

        logger.info(f"🔐 [change-password] Пароль пользователя {user.username} обновлён")
        return success_response({}, "Пароль успешно обновлён")

    except Exception as e:
        logger.error(f"[PATCH /change-password] Ошибка: {e}")
        return internal_error_response("Ошибка при смене пароля")


@router.delete("/delete-profile")
def delete_profile(db: Session = Depends(get_db), user: User = Depends(get_current_user)):
    try:
        db.delete(user)
        db.commit()
        return deleted_response("Профиль успешно удален")
    except Exception as e:
        logger.error(f"[DELETE /delete-profile] Ошибка: {e}")
        return internal_error_response("Не удалось удалить профиль")
